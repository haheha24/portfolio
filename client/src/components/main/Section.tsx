import { ReactNode, useState, useEffect, useRef } from "react";
import _ from "lodash/debounce";
import "./main.css";

interface ISection {
  setId: string;
  setTitle: string;
  children?: ReactNode;
}

const Section = ({ setId, setTitle, children = "" }: ISection) => {
  const [animation, setAnimation] = useState(false);
  const [isElementVisible, setIsElementVisible] = useState(false);
  const sectionRef = useRef<HTMLElement>(null);

  //handleScroll function to alter state
  useEffect(() => {
    const currentRef = sectionRef.current;
    //sets the isHome
    //callback to set the state of isElementVisible. If the bottom of currentRef is greater than the scrollY, set true else false
    const handleScroll = _(() => {
      window.scrollY > currentRef!.getBoundingClientRect().top
        ? setIsElementVisible(true)
        : setIsElementVisible(false);
    }, 100);

    //return useEffect early if animation has happened once already
    if (animation === true)
      return () => window.removeEventListener("scroll", handleScroll);
    //Otherwise begin handleScroll eventlistener. Start section opacity 0 for smooth transition
    if (isElementVisible === false)
      window.addEventListener("scroll", handleScroll);
    currentRef!.style.opacity = "0";
    return () => window.removeEventListener("scroll", handleScroll);
  }, [animation, isElementVisible]);

  //Once element is visible in scroll, begin animation, return opacity to normal
  useEffect(() => {
    if (isElementVisible === true) {
      sectionRef.current!.style.animationName = "section";
      sectionRef.current!.style.opacity = "1";
      //This will ensure the handleScroll useEffect will return early.
      setAnimation(true);
    }
  }, [isElementVisible]);

  return (
    <section id={setId} className="section-container" ref={sectionRef}>
      <h2 className="section-title">{setTitle}</h2>
      {children}
    </section>
  );
};

export default Section;
