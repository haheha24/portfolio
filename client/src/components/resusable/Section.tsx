import { ReactNode, forwardRef } from "react";

//Component Props
interface ISection {
  setId?: string;
  setTitle?: string;
  children?: ReactNode;
  sectionClass?: string;
  sectionStyles?: { [attribute: string]: string | number };
  headingClass?: string;
  headingStyles?: { [attribute: string]: string | number };
}

const Section = forwardRef<HTMLElement, ISection>((props: ISection, ref?) => {
  const {
    setId,
    setTitle,
    children,
    sectionClass = undefined,
    sectionStyles = undefined,
    headingClass = undefined,
    headingStyles = undefined,
  } = props;

  return (
    <section
      id={setId}
      className={sectionClass}
      style={sectionStyles}
      ref={ref}
    >
      <h2 className={headingClass} style={headingStyles}>
        {setTitle}
      </h2>
      {children}
    </section>
  );
});

export default Section;
