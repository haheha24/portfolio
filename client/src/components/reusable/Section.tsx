import { ReactNode, forwardRef } from "react";

//Component Props
interface ISection {
  setId?: string;
  setTitle?: string;
  children?: ReactNode;
  sectionClass?: string;
  sectionStyles?: { [attribute: string]: string | number };
  headingClass?: string;
  headingStyles?: { [attribute: string]: string | number };
}

const Section = forwardRef<HTMLElement, ISection>((props: ISection, ref?) => {
  const {
    setId,
    setTitle,
    children,
    sectionClass = undefined,
    sectionStyles = undefined,
    headingClass = undefined,
    headingStyles = undefined,
  } = props;

  return (
    <section
      id={setId}
      className={sectionClass}
      style={sectionStyles}
      ref={ref}
      data-testid="section-container"
    >
      <h2 className={headingClass} style={headingStyles} data-testid="section-h2" >
        {setTitle}
      </h2>
      {children}
    </section>
  );
});

export default Section;
