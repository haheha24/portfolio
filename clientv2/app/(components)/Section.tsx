import { ReactNode } from "react";
import styled from "styled-components";
import { HTMLMotionProps, motion } from "framer-motion";
import { StyledSection, StyledH1 } from "../(styles)/globalStyledComponents";

//Component Props
interface ISection {
  id?: string;
  title?: string;
  props?: HTMLMotionProps<"section">;
  children?: ReactNode;
}

//Styled and Motion Components
const SectionContainer = motion(StyledSection);
const H1 = styled(StyledH1)`
  text-align: center;
  color: ${(props) => props.theme.section.title.fontColor};
  font-size: ${(props) => props.theme.font.font40};
  width: 100%;
  margin-bottom: 30px;
  padding: 10px 0;
  background-color: ${(props) => props.theme.section.title.backgroundColor};
  box-shadow: 0 10px 10px ${(props) => props.theme.section.title.boxShadow};
`;

const Divider = styled.div`
  color: ${(props) => props.theme.section.title.fontColor};
  width: 100%;
  margin-bottom: 30px;
  padding: 10px 0;
  background-color: ${(props) => props.theme.section.title.backgroundColor};
  box-shadow: 0 10px 10px ${(props) => props.theme.section.title.boxShadow};
`;

const Section = ({ id, title, children, props }: ISection) => {
  return (
    <SectionContainer id={id} data-testid="section-container" {...props}>
      {title && <H1 data-testid="section-h2">{title}</H1>}
      {!title && <Divider data-testid="section-h2" />}
      {children}
    </SectionContainer>
  );
};

export default Section;
