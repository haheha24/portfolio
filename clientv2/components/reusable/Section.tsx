import { ReactNode } from "react";
import styled from "styled-components";
import { HTMLMotionProps, motion } from "framer-motion";
import { StyledSection, StyledH2 } from "../styledComponents";

//Component Props
interface ISection {
  id?: string;
  title?: string;
  props?: HTMLMotionProps<"section">;
  children?: ReactNode;
}

//Styled and Motion Components
const SecctionContainer = motion(StyledSection);
const H2 = styled(StyledH2)`
    text-align: center;
    color: ${(props) => props.theme.section.title.fontColor};
    font-size: 2.5rem;
    width: 100%;
    margin-bottom: 30px;
    padding: 10px 0;
    background-color: ${(props) => props.theme.section.title.backgroundColor};
    box-shadow: 0 10px 10px ${(props) => props.theme.section.title.boxShadow};
  `;

const Section = ({ id, title, children, props }: ISection) => {
  return (
    <SecctionContainer
      id={id}
      initial={{ opacity: 0, y: "250px" }}
      whileInView={{ opacity: 1, y: 0 }}
      viewport={{ once: true }}
      transition={{
        y: { duration: 1.5 },
        default: { ease: "linear" },
      }}
      data-testid="section-container"
      {...props}
    >
      <H2 data-testid="section-h2">{title}</H2>
      {children}
    </SecctionContainer>
  );
};

export default Section;
