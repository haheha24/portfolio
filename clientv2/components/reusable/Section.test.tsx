import { render, cleanup } from "@testing-library/react";
import Section from "./Section";

afterEach(() => {
  cleanup();
});

describe("Tests for the Section component", () => {
  it("Contains a section and h2 element", () => {
    const { getByTestId } = render(<Section />);
    const sectionContainer = getByTestId("section-container");
    const h2Container = getByTestId("section-h2");
    expect(sectionContainer).toBeInTheDocument();
    expect(h2Container).toBeInTheDocument();
  });

  it("renders a p tag with text", () => {
    const { container, getByTestId } = render(
      <Section>{<p data-testid="section-p">rendered a paragraph</p>}</Section>
    );
    expect(container).toContainElement(getByTestId("section-p"));
    expect(container).toHaveTextContent("rendered a paragraph");
    expect(container).not.toHaveTextContent("");
  });

  it("Accepts the id and title props", () => {
    const { container, getByTestId } = render(
      <Section title="Testing Title" id="test" />
    );
    expect(getByTestId("section-container")).toHaveAttribute("id", "test");
    expect(getByTestId("section-container")).toHaveClass("section-class");
    expect(getByTestId("section-container")).toHaveStyle("width: 50px");

    expect(getByTestId("section-h2")).toHaveClass("heading-class");
    expect(getByTestId("section-h2")).toHaveStyle("width: 50px");
    expect(container).toHaveTextContent("Testing Title");
  });
});
