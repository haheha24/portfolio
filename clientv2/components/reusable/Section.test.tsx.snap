// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the Section component renders the section component 1`] = `
.c1 {
  color: #fff;
  background-color: hsl(216,7%,25%);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 30px;
}

.c2 {
  text-align: center;
  color: #fff;
  font-size: 2.5rem;
  width: 100%;
  margin-bottom: 30px;
  padding: 10px 0;
  background-color: hsl(260,100%,60%);
  box-shadow: 0 10px 10px hsl(216,7%,14%);
}

<section
  class="c0"
  data-testid="section-container"
  style="opacity: 0; transform: translateY(250px) translateZ(0);"
>
  <h2
    class="c1 c2"
    data-testid="section-h2"
  />
</section>
`;
